pkg/build/builder/daemonless.go

```diff
@@ -311,118 +328,233 @@ func buildDaemonlessImage(sc types.SystemContext, store storage.Store, isolation
 	return err
 }
 
-func generateTransientMounts() []string {
-	mounts := []string{}
-	mounts = appendRHSMMount(defaultMountStart, mounts)
-	mounts = appendETCPKIMount(defaultMountStart, mounts)
-	mounts = appendRHRepoMount(defaultMountStart, mounts)
-	mounts = appendCATrustMount(mounts)
+type TransientMount struct {
+	buildapiv1.BuildVolumeMount
+	MountPathSource string
+	// additional buildah mount options not exposed to the Build API
+	NoDev  bool
+	NoExec bool
+	NoSuid bool
+}
+
+func (tm TransientMount) String() string {
+	var mountOptions []string
+	if !tm.BuildVolumeMount.ReadOnly {
+		mountOptions = append(mountOptions, "rw")
+	}
+	if tm.NoDev {
+		mountOptions = append(mountOptions, "nodev")
+	}
+	if tm.NoExec {
+		mountOptions = append(mountOptions, "noexec")
+	}
+	if tm.NoSuid {
+		mountOptions = append(mountOptions, "nosuid")
+	}
+	return fmt.Sprintf("%s:%s:%s", tm.MountPathSource, tm.MountPath, strings.Join(mountOptions, ","))
+
+}
+
+type TransientMounts map[string]TransientMount
+
+func (t TransientMounts) append(mount TransientMount) error {
+	if _, ok := t[mount.MountPath]; ok {
+		return fmt.Errorf("transient mount %q already exists", mount.MountPath)
+	}
+
+	t[mount.MountPath] = mount
+
+	return nil
+}
+
+func (t TransientMounts) asSlice() []string {
+	var mounts []string
+	for _, m := range t {
+		mounts = append(mounts, m.String())
+	}
 	return mounts
 }
 
-func appendRHRepoMount(pathStart string, mounts []string) []string {
+func generateTransientMounts() ([]string, error) {
+	mountsMap := make(TransientMounts)
+
+	if err := appendRHSMMount(defaultMountStart, &mountsMap); err != nil {
+		return []string{}, err
+	}
+	if err := appendETCPKIMount(defaultMountStart, &mountsMap); err != nil {
+		return []string{}, err
+	}
+	if err := appendRHRepoMount(defaultMountStart, &mountsMap); err != nil {
+		return []string{}, err
+	}
+	if err := appendCATrustMount(&mountsMap); err != nil {
+		return []string{}, err
+	}
+	if err := appendBuildVolumeMounts(&mountsMap); err != nil {
+		return []string{}, err
+	}
+	return mountsMap.asSlice(), nil
+}
+
+func appendBuildVolumeMounts(mountsMap *TransientMounts) error {
+	build := &buildapiv1.Build{}
+	if err := buildutil.GetBuildFromEnv(build); err != nil {
+		return err
+	}
+
+	var buildVolumeMounts []buildapiv1.BuildVolumeMount
+	switch {
+	case build.Spec.Strategy.Type == buildapiv1.SourceBuildStrategyType:
+		buildVolumeMounts = build.Spec.Strategy.SourceStrategy.VolumeMounts
+	case build.Spec.Strategy.Type == buildapiv1.DockerBuildStrategyType:
+		buildVolumeMounts = build.Spec.Strategy.DockerStrategy.VolumeMounts
+	}
+
+	for _, bvm := range buildVolumeMounts {
+		var mountOptions []string
+		if !bvm.ReadOnly {
+			mountOptions = append(mountOptions, "rw")
+		} else {
+			mountOptions = append(mountOptions, "rw")
+		}
+
+		if err := mountsMap.append(TransientMount{
+			BuildVolumeMount: bvm,
+			MountPathSource:  filepath.Join(BuildVolumeMountPath, bvm.MountPath),
```

Comment:
Now I see what you are doing in openshift-controller-manager - you are laying the mounts out based on their destination paths. It helps alleviate the volume name collision problem. `filepath.Join` also strips the leading `/` if we are given an absolute file path, which is good. We should probably convert relative file paths to an absolute filepath within buildah's runtime environment.